### PR TITLE
Created TableUtil containing former static Table functions

### PR DESCRIPTION
**Pull request by gervandiepen**
Class Table contained several static functions which have been moved to the new TableUtil. A few more functions have been added to deal with creating  a new table and handling subtables.
Several casacore classes have been adapted for this change.

This is part of issue #1143.

### CODE REVIEW

File: tables/Tables/TableUtil.cc [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+        // Check if main table exists.
+        if (! Table::isReadable (names[0])) {
+          msg = "main table " + names[0] + " does not exist";
+        } else {
+          tab = Table(names[0]);
+          // Get name of last subtable.
+          if (names.size() > 1) {
+            lastPart = names[names.size()-1];
+            // Check if all subtables exist, maybe except last one.
+            uInt nrname = mustExist ? names.size() : names.size()-1;
```
Comment (mnijhuis-tos) [inline tables/Tables/TableUtil.cc:302]: I'm somewhat confused what the return value of `tryFindSubTable` is in different cases: The value of `tab` may vary:
- If `mustExist` is true, `tab` is always the last table, with name `lastPart`.
- If `mustExist` is false, `tab` is the last-but-one table, and `lastPart` is the name of a subtable of `tab` that may or may not exist.

Maybe it's simpler if `tab` is always the last-but-one table: The semantics of the return value are then always the same. At the end of the function, you can then add `if (mustExist && !tab.keywordSet().isDefined(lastPart)) { msg = ...; }`

Also, please clearly document the return value in TableUtil.h.

File: tables/Tables/test/tTableUtil.cc [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+//# Correspondence concerning AIPS++ should be addressed as follows:
+//#        Internet email: aips2-request@nrao.edu.
+//#        Postal address: AIPS++ Project Office
+//#                        National Radio Astronomy Observatory
+//#                        520 Edgemont Road
+//#                        Charlottesville, VA 22903-2475 USA
+//#
+//# $Id$
+
+//# Includes
```
Comment (mnijhuis-tos) [inline tables/Tables/test/tTableUtil.cc:28]: Again, please remove.

File: tables/Tables/test/tTableUtil.cc [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+// This program tests some table utility functions.
+
+
+// Create a table and various subtables.
+void createTables()
+{
+  // Create a main table.
+  TableDesc td("", "", TableDesc::Scratch);
+  td.addColumn (ScalarColumnDesc<Int>("RowNr"));
+  Table tab(TableUtil::createTable("tTableUtil_tmp/maindata", td, Table::New));
```
Comment (mnijhuis-tos) [inline tables/Tables/test/tTableUtil.cc:50]: Use a `const std::string` or `const String` that contains the magic constant "tTableUtil_tmp/maindata". You can then use it throughout this file.

File: tables/Tables/test/tTableUtil.cc [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+    testDelete();
+    testReplace();
+    testErrors();
+    testLayoutInfo("tTableUtil_tmp/maindata");
+    testLayoutInfo("tTableUtil_tmp/maindata::SubTab2");
+  } catch (const std::exception& x) {
+    cout << "Caught an exception : " << x.what() << endl;
+    return 1;
+  } 
+  cout << "OK" << endl;
```
Comment (mnijhuis-tos) [inline tables/Tables/test/tTableUtil.cc:273]: I rather see something more explicit, since anything could have written "OK".
```suggestion
  cout << "tTableUtil: All tests passed." << endl;
```

File: tables/Tables/test/tTableUtil.cc [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+#include <casacore/tables/Tables/TableUtil.h>
+#include <casacore/tables/Tables/SetupNewTab.h>
+#include <casacore/tables/Tables/TableRecord.h>
+#include <casacore/tables/Tables/TableDesc.h>
+#include <casacore/tables/Tables/ScaColDesc.h>
+#include <casacore/casa/Utilities/Assert.h>
+#include <casacore/casa/OS/Directory.h>
+#include <casacore/casa/iostream.h>
+
+#include <casacore/casa/namespace.h>
```
Comment (mnijhuis-tos) [inline tables/Tables/test/tTableUtil.cc:39]: [Google forbids importing an entire namespace.](https://google.github.io/styleguide/cppguide.html#Namespaces) Instead, only use what you need, for example:
```
using casacore::Table;
using casacore::TableDesc;
using casacore::TableUtil;
```

File: tables/Tables/Table.cc [hunk 6]
```diff
@@ -362,6 +363,7 @@ Table& Table::operator= (const Table& that)
     return *this;
 }
 
+/*
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/Tables/Table.cc:366]: This was an oversight. Removed it.

File: tables/Tables/TableUtil.cc [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+                            const String& subtableName,
+                            Bool checkSubTables)
+    {
+      // Get the full table name of the subtable.
+      // Make sure the Table object is deleted, otherwise isOpened is always true.
+      String fullName;
+      {
+        Table table = parent.keywordSet().asTable (subtableName);
+        fullName = table.tableName();
+      }
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/Tables/TableUtil.cc:176]: Is it guaranteed that the temporary Table object is deleted before the next line?

File: tables/Tables/TableUtil.cc [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+               ("TableUtil::getLayout: unknown table kind " + tp));
+      }
+      ios.close();
+      return nrow;
+    }
+
+    TableInfo tableInfo (const String& tableName)
+      {
+        return BaseTable::tableInfo (getFullName(tableName));
+      }
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/Tables/TableUtil.cc:265]: I've replaced the tabs by spaces and re-indented all code.
An automatic tool would generate a massive change. Furthermore, different casacore developers usually like their own style and I think it would be hard to find a compromise.

File: tables/Tables/TableUtil.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (mnijhuis-tos) [reply tables/Tables/TableUtil.cc:176]: Yes. From the C++ standard:

12.2 Temporary objects, clause 3: "Temporary objects are destroyed as the last step in evaluating the full-expression (1.9) that (lexically) contains the point where they were created."

1.9 Program execution, clause 12: "A full-expression is an expression that is not a subexpression of another expression."

File: tables/Tables/Table.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/Tables/Table.cc:366]: It was an oversight that the outcommented code was still present. Removed it.

File: tables/Tables/TableUtil.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/Tables/TableUtil.cc:176]: Is it guaranteed that the temporary Table object is removed before the next line is executed or has the compiler the freedom to delete it only at the end of the function?

File: tables/Tables/TableUtil.cc [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+        // No subtable given, so create a main table.
+        SetupNewTable newtab(tableName, desc,
+                             tableName.empty() ? Table::Scratch : tabOpt,
+                             storageOption);
+        newtab.bindCreate (dmInfo);
+        // Create the table.
+        return Table(newtab, tabType, lockOptions, nrrow, initialize, endian, tsmOpt);
+      }
+      return createSubTable (t.first, t.second, desc, tabOpt, storageOption,
+                             dmInfo, lockOptions, nrrow, initialize, endian, tsmOpt);
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/Tables/TableUtil.cc:91]: In the past compilers used to complain about 'no return' if put in an else branch. Probably they are smarter now.
But what is the advantage?

File: tables/Tables/TableUtil.cc [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+                     Table::TableOption option,
+                     const TSMOption& tsmOption)
+    {
+      // See if the table can be opened as such.
+      if (Table::isReadable(tableName)) {
+        return Table(tableName, lockOptions, option, tsmOption);
+      }
+      // Try to find and open the last subtable by splitting at ::
+      // It throws an exception if not possible.
+      std::pair<Table,String> t = tryFindSubTable (tableName, True);
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/Tables/TableUtil.cc:61]: Fixed by tryFindSubTable changes.

File: tables/Tables/TableUtil.cc [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+//# Correspondence concerning AIPS++ should be addressed as follows:
+//#        Internet email: aips2-request@nrao.edu.
+//#        Postal address: AIPS++ Project Office
+//#                        National Radio Astronomy Observatory
+//#                        520 Edgemont Road
+//#                        Charlottesville, VA 22903-2475 USA
+//#
+//# $Id$
+
+//# Includes
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/Tables/TableUtil.cc:28]: $Id$ is part of standard casacore header, although it is old cvs/svn functionality.

File: tables/Tables/TableUtil.cc [hunk 12]
```diff
@@ -0,0 +1,321 @@
...
+      } else if (names.size() > 1  ||  mustExist) {
+        // No need to check if new table and no subtable given.
+        // Check if main table exists.
+        if (! Table::isReadable (names[0])) {
+          msg = "main table " + names[0] + " does not exist";
+        } else {
+          tab = Table(names[0]);
+          // Get name of last subtable.
+          if (names.size() > 1) {
+            lastPart = names[names.size()-1];
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/Tables/TableUtil.cc:300]: Changed

File: tables/Tables/TableUtil.cc (hunk 1 above)
Comment (gervandiepen) [reply tables/Tables/TableUtil.cc:302]: Changed name to findParentTable and removed Bool argument.
openTable now does the last open, which also solves problem that LockOption was not used.

File: tables/Tables/test/tTableUtil.cc (hunk 3 above)
Comment (gervandiepen) [reply tables/Tables/test/tTableUtil.cc:50]: Done

File: tables/Tables/test/tTableUtil.cc (hunk 5 above)
Comment (gervandiepen) [reply tables/Tables/test/tTableUtil.cc:39]: Casacore forbids imports of class, function or namespace in header files, so anybody can safely uses them. But imports are allowed it in .cc files because that scope is limited.
I've changed the include into a normal using statement.

File: tables/Tables/TableUtil.cc (hunk 12 above)
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/Tables/TableUtil.cc:300]: Undid the change, because it changed the behaviour in case names.size()==1.

File: tables/Tables/TableUtil.cc [hunk 13]
```diff
@@ -0,0 +1,321 @@
...
+      return tabName;
+    }
+    
+    // Return Table and name of last part.
+    std::pair<Table,String> tryFindSubTable (const String& fullName, Bool mustExist)
+    {
+      Table tab;
+      String lastPart, msg;
+      // Split the name on :: to get the main and subtable names.
+      Vector<String> names = stringToVector(fullName, std::regex("::"));
```
[the remark this answers is not in the document]
Comment (gervandiepen) [reply tables/Tables/TableUtil.cc:286]: changed

File: tables/Tables/TableUtil.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (mnijhuis-tos) [reply tables/Tables/TableUtil.cc:98]: Using `else` makes the code more explicit and a therefore a little more readable. I think compilers shouldn't complain about missing returns anymore in such cases.

However, I said this improvement is 'Optional', so feel free to ignore it.

File: tables/Tables/Table.h [hunk 14]
```diff
@@ -1003,33 +1003,33 @@ friend class TableExprNodeRep;
     // Some deprecated functions for backward compatibility, now in TableUtil.h.
     // Use old way of indicating deprecate to avoid -Wc++14-extensions warnings.
     // <group>
-     // [[deprecated ("Now use TableUtil::openTable")]]
+     //# [[deprecated ("Now use TableUtil::openTable")]]
```
Comment (mnijhuis-tos) [inline tables/Tables/Table.h:1006]: Why did you add the # ? I don't mind, but I couldn't find any documentation what # does here.

File: tables/Tables/TableUtil.cc [hunk 15]
```diff
@@ -57,9 +57,16 @@ namespace casacore {
...
+      std::pair<Table,String> t(findParentTable(tableName));
+      if (t.first.isNull()) {
+        throw TableError ("Table " + tableName + " does not exist");
+      }
+      // Find the last subtable in the parent.
+      if (! t.first.keywordSet().isDefined(t.second)) {
+        throw TableError ("Table name " + tableName + " is invalid (" +
+                          "subtable " + t.second + " is unknown)");
+      }
+      return t.first.keywordSet().asTable (t.second);
```
Comment (mnijhuis-tos) [inline tables/Tables/TableUtil.cc:69]: You still don't use `lockOptions`, `option` and `tsmOption` when opening a subtable.

File: tables/Tables/TableUtil.cc [hunk 16]
```diff
@@ -186,8 +190,8 @@ namespace casacore {
       }
       // See if the name represents a table.
       if (! Table::isReadable(tableName)) {
-        // See if the name contains subtable names using ::
-        std::pair<Table,String> t = tryFindSubTable (tableName, False);
+        // See< if the name contains subtable names using ::
```
Comment (mnijhuis-tos) [inline tables/Tables/TableUtil.cc:193]: Remove '<'

File: tables/Tables/TableUtil.cc [hunk 17]
```diff
@@ -186,8 +190,8 @@ namespace casacore {
...
+        std::pair<Table,String> t = findParentTable (tableName);
         if (! (t.first.isNull()  ||  t.second.empty())) {
```
Comment (mnijhuis-tos) [inline tables/Tables/TableUtil.cc:195]: Only the first part of this check should be enough:
```
if (!t.first.isNull()) {
```

File: tables/Tables/TableUtil.cc [hunk 18]
```diff
@@ -137,28 +144,28 @@ namespace casacore {
                          Bool checkSubTables, Bool splitColons)
     {
       if (splitColons) {
-        std::pair<Table,String> t = tryFindSubTable (tableName, True);
+        std::pair<Table,String> t = findParentTable (tableName);
         if (! (t.first.isNull()  ||  t.second.empty())) {
```
Comment (mnijhuis-tos) [inline tables/Tables/TableUtil.cc:148]: Only checking `!t.first.isNull()` should be enough.

File: tables/Tables/TableUtil.cc [hunk 19]
```diff
@@ -260,15 +264,15 @@ namespace casacore {
...
+      return BaseTable::tableInfo (getFullName(tableName));
+    }
 
     String getFullName (const String& tableName)
     {
       // See if a subtable is given using ::.
       String tabName;
-      std::pair<Table,String> t = tryFindSubTable (tableName, False);
+      std::pair<Table,String> t = findParentTable (tableName);
       if (t.first.isNull()  ||  t.second.empty()) {
```
Comment (mnijhuis-tos) [inline tables/Tables/TableUtil.cc:276]: `t.first.isNull()` by itself already says if a subtable is given.

File: tables/Tables/TableUtil.cc [hunk 20]
```diff
@@ -288,26 +292,23 @@ namespace casacore {
...
-                tab = Table();
-                break;
-              }
-              tab = tab.keywordSet().asTable (names[i]);
+          lastPart = names[names.size()-1];
+          // Check if all subtables exist, except last one.
+          for (uInt i=1; i<names.size()-1; ++i) {
+            if (! tab.keywordSet().isDefined(names[i])) {
+              msg = "subtable " + names[i] + " is unknown";
+              tab = Table();
```
Comment (mnijhuis-tos) [inline tables/Tables/TableUtil.cc:308]: Optional: Resetting `tab` isn't strictly necessary, since you're throwing an exception below. If it is necessary, I would reset `tab` just before throwing the exception, so it's always reset in all error cases.

File: tables/Tables/test/tTableUtil.cc (hunk 5 above)
Comment (mnijhuis-tos) [reply tables/Tables/test/tTableUtil.cc]: For .cc files, Google also forbids importing an entire namespace. I don't know to what extend we want to follow Google style for Casacore, though.

File: tables/Tables/TableUtil.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (mnijhuis-tos) [reply tables/Tables/TableUtil.cc:28]: I only meant the `//# Includes`

File: tables/Tables/TableUtil.cc (hunk 13 above)
[the remark this answers is not in the document]
Comment (mnijhuis-tos) [reply tables/Tables/TableUtil.cc:290]: GitHub doesn't show any change...

**Review by mnijhuis-tos — CHANGES_REQUESTED** [review]

File: tables/Tables/Table.h (hunk 14 above)
Comment (gervandiepen) [reply tables/Tables/Table.h:1006]: //# lines are not included in the doxygen documentation.
casacore used to have its own documentation processor, but switched to doxygen quite some time ago. A preprocessing script turns comment lines with casacore tags into doxygen lines.

File: tables/Tables/TableUtil.cc (hunk 18 above)
Comment (gervandiepen) [reply tables/Tables/TableUtil.cc:148]: Your're right. Changed at all places.

File: tables/Tables/TableUtil.cc (hunk 15 above)
Comment (gervandiepen) [reply tables/Tables/TableUtil.cc:70]: Added them to findParentTable.

File: tables/Tables/TableUtil.cc (hunk 16 above)
Comment (gervandiepen) [reply tables/Tables/TableUtil.cc]: Done

**Review by mnijhuis-tos — APPROVED** [review]
Since there's only one minor unresolved conversation left, I'll approve.
